### ISSUE TITLE
[Python Package] Float point exception: 8 when running predict or predict_proba

### ISSUE THREAD

**Issue by JovanVeljanoski** [Microsoft/LightGBM#1174]
Hi, 
I run  the latest version of LightGBM (installed from source) recently and I am encountering a problem I have not met before. I am training a binary classifier on a very sparse dataset (scipy.sparse.csr.csr_matrix) of shape (64220, 759414), and validating on a similar matrix of shape  (31631, 759414).  Both sets fit well into the memory of my computer. The training goes well at first and I am reaching an acceptable train and validation loss. 

However, when I try to get the predictions or probabilities for the validation set, the kernel of the jupyter notebook dies, or if I run the same in a shell it gives "Float point exception: 8". The same happens if use the sklearn API, or the LightGBM .train API. 

I also notice that, while training, I get a message such as:

[LightGBM] [Info] Number of positive: 6217, number of negative: 58003
[LightGBM] [Info] Total Bins 467101
[LightGBM] [Info] Number of data: 64220, number of used features: 15292
[LightGBM] [Warning] Accuracy may be bad since you didn't set num_leaves and 2^max_depth > num_leaves.
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf

Which is weird since I do pass on parameters given which 2^max_depth > num_leaves. 
It is the 1st time that I am using a dataset which is sparse and having so many columns, so perhaps I am not doing something which I should, which results in the issues I am getting. 
I also don't understand why one of the warnings states that only 15292 features are used when the sparse matrix is of a different shape. Could this be why kernel is crashing? Is there a parameter I am not seeing that will allow for more (all?) features to be used, or is this a limitation for now?

I could not reproduce this issue with any of the other (smaller and dense) datasets that I have at hand at the moment.

I am running this all on a MacBook Pro, OS X 10.12.6 (16GB RAM). I have python 3.6 installed through conda, and LightGBM is pulled from github, and compiled and installed locally.  

Any help or advise would be highly appreciated. Thank you.
Jovan.

**Comment by guolinke** [comment]
can you share a example code ? 
For the number of features, LightGBM will auto remove the features with almost all zeros. So it isn't the problems. If you really want to use these almost-all-zero features, you can try to set `min_data=1`. 

For the `2^max_depth > num_leaves`, I think it is caused the wrong parameters.  

For the `Float point exception: 8`, we didn't see it before. Maybe it is some bugs on OS X. Can you try to run on windows/linux machines ?

**Comment by guolinke** [comment]
BTW, the `Float point exception: 8` is thrown in jupyter notebook when running prediction? and the normal training in native python is good ?

**Comment by JovanVeljanoski** [comment]
Hi,
Thank you for the fast response, really appreciate it. While training, everything seems to go well in both the jupyter notebook and in native python. In both cases I get the loss from the train and validation sets printed to disk, and it makes sense. 

The problem happens when I try to predict. In the jupyter notebook the kernel dies, and in the native python shell the float point exception is thrown.  I try to predict on the train or the validation set that I just used for the training phase.

I will try to reproduce this with some random data. I am happy to share the data that I have now, it is not sensitive, but it is large (so not practical).. Below is the code that i am using (so you can see how I train, and which params I use).

Do you see anything wrong here? 
Thank you!


```
import scipy.sparse
import pandas as pd
import lightgbm as lgbm

X_vec = scipy.sparse.load_npz('./1st_try/X_vec.npz')
V_vec = scipy.sparse.load_npz('./1st_try/V_vec.npz')
Y_vec = scipy.sparse.load_npz('./1st_try/Y_vec.npz')
yx = np.load('yx.npy')
yv = np.load('yv.npy')


dtrain = lgbm.Dataset(data=X_vec, label=yx[:,0])
dval   = lgbm.Dataset(data=V_vec, label=yv[:,0])

params = {
    'max_depth': 7,
    'learning_rate': 0.05,
    'num_leaves': 31,
    'application': 'binary',
    'metric': 'binary_logloss',
    'min_data_in_leaf': 20,
    'subsample': 0.8,
    'colsample_bytree': 0.8,
    'reg_lambda': 0.5,
    'reg_alpha': 1.5,
}

h0 = {}

booster = lgbm.train(params=params, 
                     train_set=dtrain, 
                     num_boost_round=10, 
                     valid_sets=[dtrain,dval],
                     valid_names=['train','val'],
                     evals_result=h0,
                     early_stopping_rounds=50,
                     verbose_eval=True)

print('Predict')
tmp = booster.predict(V_vec)
print(tmp)
```

**Comment by JovanVeljanoski** [comment]
Small update:

By creating a sparse matrix of similar shape does not re-create the problem. But with the dataset in question, if I just reduce the number of features used to <10000, the problem disappears, and I can train and predict without any errors. 

Setting `min_data=1` does not help in this case.

**Comment by guolinke** [comment]
@JovanVeljanoski `num_leaves=31` is default value, so it will give a warning. You can set it to another value to avoid it.
For the test, What is dimension for your test data when crashed? Is it CSR format?

**Comment by JovanVeljanoski** [comment]
Hi, 

Yes the format is CSR. I train data has a shape of (64220, 759414), and validation data has a shape of (31631, 759414).  I am trying to run the predict on the validation set (although the train set gives the same problem). I also tried predicting on just a few rows (10 or so), and the same problem happens again.

Changing `num_leaves`(to a higher value) did not matter.

**Comment by guolinke** [comment]
@JovanVeljanoski 
It is very strange. A reproduce example will be better.

**Comment by ktleung2017** [comment]
I reproduced the problem with the following code:

```
import numpy as np
import scipy.sparse
import lightgbm as lgbm


n_feature = 100000

X_train = scipy.sparse.rand(10000, n_feature)
X_cv = scipy.sparse.rand(5000, n_feature)
y_train = np.random.randint(2, size=10000)
y_cv = np.random.randint(2, size=5000)

lgb_train = lgbm.Dataset(data=X_train, label=y_train)
lgb_cv = lgbm.Dataset(data=X_cv, label=y_cv, reference=lgb_train)

params = {
    'max_depth': 7,
    'learning_rate': 0.05,
    'num_leaves': 31,
    'application': 'binary',
    'metric': 'binary_logloss',
    'min_data_in_leaf': 20,
    'subsample': 0.8,
    'colsample_bytree': 0.8,
    'reg_lambda': 0.5,
    'reg_alpha': 1.5,
}

h0 = {}

booster = lgbm.train(params=params,
                     train_set=lgb_train,
                     num_boost_round=10,
                     evals_result=h0,
                     valid_sets=lgb_cv,
                     early_stopping_rounds=50,
                     verbose_eval=True)

# This works
booster.predict(scipy.sparse.csr_matrix(np.ones(n_feature), shape=(1, n_feature)))

# This works
booster.predict(scipy.sparse.csr_matrix(np.random.rand(n_feature), shape=(1, n_feature)))

# This fails
booster.predict(scipy.sparse.csr_matrix(np.zeros(n_feature), shape=(1, n_feature)))
```

Looks like if we try to `predict` a model with large number of features by a zero vector, it will throw exception.

Note:
- when `n_feature` is set to `100`, no exceptions will be thrown
- behavior of this problem in other classes (`csc_matrix`, `ndarray`) has not been investigated. I only looked at `csr_matrix`.

**Comment by guolinke** [comment]
@JovanVeljanoski can you try the latest code ?

**Comment by JovanVeljanoski** [comment]
Hi,

The latest code works as expected. Thank you very much @guolinke @ktleung2017  for your prompt action on this.

Cheers,
Jovan.

**Comment by felixcra** [comment]
I just encountered this issue too. I think it is very strange behaviour, that there an error message is shown when using the default value, which disappears if increasing or decreasing `num_leafes` by one.
